core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java

```diff
@@ -71,49 +74,109 @@ public Endpoint select(ClientRequestContext ctx) {
-            private final int totalWeight;
+            private final long totalWeight; // prevent overflow by using long
+
+            private static final class EndpointsGroupByWeight {
+                int startIndex;
+                int weight;
+                long accumulatedWeight;
+
+                EndpointsGroupByWeight(int startIndex, int weight, long accumulatedWeight) {
+                    this.startIndex = startIndex;
+                    this.weight = weight;
+                    this.accumulatedWeight = accumulatedWeight;
+                }
+            }
+
+            private final EndpointsGroupByWeight[] endpointsGroupByWeight;
 
             EndpointsAndWeights(Iterable<Endpoint> endpoints) {
                 int minWeight = Integer.MAX_VALUE;
                 int maxWeight = Integer.MIN_VALUE;
-                int totalWeight = 0;
+                long totalWeight = 0;
+
+                // get min and max weight
                 for (Endpoint endpoint : endpoints) {
                     final int weight = endpoint.weight();
                     minWeight = Math.min(minWeight, weight);
                     maxWeight = Math.max(maxWeight, weight);
-                    totalWeight += weight;
                 }
-                this.endpoints = ImmutableList.copyOf(endpoints);
-                this.maxWeight = maxWeight;
+
+                // prepare endpoints
+                List<Endpoint> endps = ImmutableList.copyOf(endpoints)
+                        .stream()
+                        .filter(endpoint -> endpoint.weight() > 0) // only process endpoint with weight > 0
+                        .sorted(Comparator
+                                .comparing(Endpoint::weight)
+                                .thenComparing(Endpoint::host)
+                                .thenComparingInt(Endpoint::port))
+                        .collect(Collectors.toList());
+                int numEndpoints = endps.size();
+
+                // accumulation
+                LinkedList<EndpointsGroupByWeight> accumulatedGroups = new LinkedList<>();
+                EndpointsGroupByWeight currentGroup = null;
+                int rest = numEndpoints;
+                for (Endpoint endpoint : endps) {
+                    if (currentGroup == null || currentGroup.weight != endpoint.weight()) {
+                        totalWeight += currentGroup == null ?
+                                (long) endpoint.weight() * (long) rest
+                                : (long) (endpoint.weight() - currentGroup.weight) * (long) rest;
+                        currentGroup = new EndpointsGroupByWeight(numEndpoints - rest,
+                                endpoint.weight(), totalWeight);
+                        accumulatedGroups.addLast(currentGroup);
+                    }
+
+                    rest--;
+                }
+
+                this.endpoints = endps;
+                this.endpointsGroupByWeight = accumulatedGroups.toArray(
```

Comment:
You don't need to convert to raw array if you use `ArrayList`, I think.

Reply:
I think ArrayList is much more better in term of abstraction and grow. 

However in this PR, I just want to make this micro optimization as good as possible, not calling additional function to archive element.

I think I would count number of distinct weight and do preallocation.

Reply:
I think using an `ArrayList` will generally be faster than traversing a `LinkedList` that happens here. It seems reasonable to just allocate `new ArrayList<(endps.size())` and call `ImmutableList.copyOf` here.

Also, since this should only be called when endpoints change, which isn't often, I don't think we need to look at overoptimizations in this constructor.